src/clj/fluree/db/query/exec/update.cljc

```diff
@@ -0,0 +1,127 @@
+(ns fluree.db.query.exec.update
+  (:require [fluree.db.flake :as flake]
+            [fluree.db.constants :as const]
+            [fluree.db.dbproto :as dbproto]
+            [fluree.db.json-ld.iri :as iri]
+            [fluree.db.datatype :as datatype]
+            [fluree.db.query.exec.where :as where]
+            [fluree.db.util.log :as log]
+            [clojure.core.async :as async]))
+
+(defn assign-clause
+  [clause solution]
+  (map (fn [triple]
+         (where/assign-matched-values triple solution))
+       clause))
+
+(defn retract-triple-matches
+  [db t fuel-tracker error-ch matched-ch]
+  (let [retract-ch (async/chan 2 (comp cat
+                                       (map (fn [f]
+                                              (flake/flip-flake f t)))))]
+    (async/pipeline-async 2
+                          retract-ch
+                          (fn [matched-triple ch]
+                            (-> db
+                                (where/resolve-flake-range fuel-tracker error-ch matched-triple)
+                                (async/pipe ch)))
+                          matched-ch)
+    retract-ch))
+
+(defn retract
+  [db txn {:keys [t] :as _tx-state} fuel-tracker error-ch solution-ch]
+  (let [clause     (:delete txn)
+        matched-ch (async/pipe solution-ch
+                               (async/chan 2 (comp (mapcat (partial assign-clause clause))
+                                                   (filter where/all-matched?)
+                                                   (map (partial where/compute-sids db))
+                                                   (remove nil?))))]
+    (retract-triple-matches db t fuel-tracker error-ch matched-ch)))
+
+(defn build-sid
+  [{:keys [namespaces] :as _db} ns nme]
+  (let [ns-code (get namespaces ns)]
+    (iri/->sid ns-code nme)))
+
+(defn generate-sid!
+  [db-vol iri]
+  (let [[ns nme] (iri/decompose iri)]
+    (-> db-vol
+        (vswap! (fn [db]
```

Comment:
Is this volatile safe? Under high contention we may end up with two different namespaces that have the same code. It may be unlikely but that would really bite when it happens.

Reply:
This volatile is only updated inside a transducer, which is exactly the use case volatiles were the clojure standard library to support.